ion/src/shared/storage.cpp

```diff
@@ -0,0 +1,278 @@
+ for (char * p : *this) {
+    Record currentRecord(nameOfRecordStarting(p));
+    if (record == currentRecord) {
+      const char * name = nameOfRecordStarting(p);
+      record_size_t size = sizeOfRecordStarting(p);
+      const void * value = valueOfRecordStarting(p);
+      return {.buffer= value, .size= size-strlen(name)-1-sizeof(record_size_t)};
+    }
+  }
+  return {.buffer= nullptr, .size= 0};
+}
+
+Storage::Record::ErrorStatus Storage::setValueOfRecord(Record record, Record::Data data) {
+  for (char * p : *this) {
+    Record currentRecord(nameOfRecordStarting(p));
+    if (record == currentRecord) {
+      record_size_t previousRecordSize = sizeOfRecordStarting(p);
+      const char * name = nameOfRecordStarting(p);
+      size_t newRecordSize = sizeOfRecord(name, data.size);
+      if (newRecordSize >= k_maxRecordSize || !slideBuffer(p+previousRecordSize, newRecordSize-previousRecordSize)) {
+        return Record::ErrorStatus::NoEnoughSpaceAvailable;
+      }
+      record_size_t nameSize = strlen(name)+1;
+      overrideSizeAtPosition(p, newRecordSize);
+      overrideValueAtPosition(p+sizeof(record_size_t)+nameSize, data.buffer, data.size);
+      return Record::ErrorStatus::None;
+    }
+  }
+  return Record::ErrorStatus::RecordDoesNotExist;
+}
+
+void Storage::destroyRecord(Record record) {
+  for (char * p : *this) {
+    Record currentRecord(nameOfRecordStarting(p));
+    if (record == currentRecord) {
+      record_size_t previousRecordSize = sizeOfRecordStarting(p);
+      slideBuffer(p+previousRecordSize, -previousRecordSize);
+    }
+  }
+}
+
+Storage::record_size_t Storage::sizeOfRecordStarting(char * start) const {
+  return *((record_size_t *)start);
+}
+
+const char * Storage::nameOfRecordStarting(char * start) const {
+  return start+sizeof(record_size_t);
+}
+
+const void * Storage::valueOfRecordStarting(char * start) const {
+  char * currentChar = start+sizeof(record_size_t);
+  while (*currentChar != 0) {
+    currentChar++;
+  }
+  return currentChar+1;
+}
+
+size_t Storage::overrideSizeAtPosition(char * position, record_size_t size) {
+  *((record_size_t *)position) = size;
+  return sizeof(record_size_t);
+}
+
+size_t Storage::overrideNameAtPosition(char * position, const char * name) {
+  return strlcpy(position, name, strlen(name)+1)+1;
+}
+
+size_t Storage::overrideValueAtPosition(char * position, const void * data, record_size_t size) {
+  memcpy(position, data, size);
+  return size;
+}
+
+bool Storage::isNameTaken(const char * name, Record * recordToExclude) {
+  Record r = Record(name);
+  if (r == Record()) {
+    return true;
+  }
+  for (char * p : *this) {
+    Record s(nameOfRecordStarting(p));
+    if (recordToExclude && s == *recordToExclude) {
+      continue;
+    }
+    if (s == r) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool Storage::nameCompliant(const char * name) const {
+  /* The name format is [a-z0-9_.]+ */
```

Comment:
That regex does not match its implementation (only one dot is allowed, empty string allowed).